keps/sig-node/277-ephemeral-containers/README.md

```diff
@@ -675,60 +663,109 @@ _This section must be completed when targeting alpha to a release._
 
+  Since this feature is not critical to production workloads, the main risk
+  is that enabling the feature will adversely affect other components when
+  it becomes enabled by default.
+
+  This feature is implemented in the apiserver and kubelet. The change to the
+  core API has had several releases to soak. For the apiserver, the main risk
+  is described by [Adding Unstable Features to Stable Versions]. Specifically:
+
+  > Ensuring existing data is preserved is needed so that when the feature is
+  > enabled by default in a future version n and data is unconditionally allowed
+  > to be persisted in the field, an n-1 API server (with the feature still
+  > disabled by default) will not drop the data on update.
+
+  We've followed the instructions in this doc for how to persist unstable fields
+  during an update. The case for this feature is slightly more complicated
+  because the field may not be set by the default update resource. This logic
+  has been in place for several releases.
+
+  The apiserver and kubelet have existing upgrade tests
+  (https://testgrid.k8s.io/google-gce-upgrade), but these likely don't exercise
+  ephemeral containers. We'll investigate whether it's possible to add ephemeral
+  containers to these existing tests.
+
+  TODO(verb): create issue to update existing upgrade tests after discussing
+  with PRR reviewer.
+
 * **Is the rollout accompanied by any deprecations and/or removals of features, APIs, 
 fields of API types, flags, etc.?**
-  Even if applying deprecation policies, they may still surprise some users.
+
+  No.
 
 ### Monitoring Requirements
 
 _This section must be completed when targeting beta graduation to a release._
 
 * **How can an operator determine if the feature is in use by workloads?**
-  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
-  checking if there are objects with field X set) may be a last resort. Avoid
-  logs or events for this purpose.
+
+  The kubelet has a gauge metric `kubelet_running_containers` that we can
+  augment to contain a count of running ephemeral containers.
+
+  TODO(verb): Open an issue to create this metric after discussing with PRR
+  reviewer.
 
 * **What are the SLIs (Service Level Indicators) an operator can use to determine 
 the health of the service?**
-  - [ ] Metrics
-    - Metric name:
-    - [Optional] Aggregation method:
-    - Components exposing the metric:
+  - [x] Metrics
+    - Metric name: `apiserver_request_total{component="apiserver",resource="pods",subresource="ephemeralcontainers"}` (apiserver), `kubelet_container_errors_total{type="Ephemeral"}` (kubelet, Proposed)
+    - [Optional] Aggregation method: Aggregate by container type
+    - Components exposing the metric: kubelet
   - [ ] Other (treat as last resort)
     - Details:
 
+  Note that the kubelet SLI for this feature is a counter that increments upon
+  failure to create an ephemeral container. Right now the kubelet only surfaces
```

Comment:
Would this have noise due to user error? That is, can we differentiate between a kubelet failure that should be of concern to the operator vs the user using a bad image or something? I guess we have the same issue for regular containers, so we should follow the same logic we have there.

Reply:
Ah, that's a good point. I agree that this "server error" vs "user error" is important here, but I'm not sure of the current behavior of the kubelet yet.  I [brought it up in the metrics issue](https://github.com/kubernetes/kubernetes/issues/97974#issuecomment-762356411) to discuss with sig-instrumentation and will make sure it's addressed.